### PR TITLE
Apply P0663R0 PR for #245 The iterator adaptors should customize iter…

### LINKED ISSUE #245: The iterator adaptors should customize iter_move and iter_swap
### Proposed resolution:

In [reverse.iterator], add declarations to the synopsis:

> ```diff
>   reference operator[](difference_type n) const
>     requires RandomAccessIterator<I>();
> +
> + friend constexpr rvalue_reference_t<I> iter_move(const reverse_iterator& i)
> +   noexcept(see below);
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const reverse_iterator& x, const reverse_iterator<I2>& y)
> +     noexcept(see below);
> +
> private:
>   I current; // exposition only
> };
> ```

Add new sections before [reverse.iter.make] as follows:

> ```diff
> + 24.7.1.3.? iter_move [reverse.iter.iter_move]
> +
> + friend constexpr rvalue_reference_t<I> iter_move(const reverse_iterator& i)
> +   noexcept(see below);
> +
> + 1 Effects: Equivalent to: return ranges::iter_move(prev(i.current));
> +
> + 2 Remarks: The expression in the noexcept clause is equivalent to 
> +   noexcept(ranges::iter_move(declval<I&>())) && noexcept(--declval<I&>()) && 
> +   is_nothrow_copy_constructible<I>::value.
> +
> + 24.7.1.3.? iter_swap [reverse.iter.iter_swap]
> +
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const reverse_iterator& x, const reverse_iterator<I2>& y)
> +     noexcept(see below);
> +
> + 1 Effects: Equivalent to ranges::iter_swap(prev(x.current), prev(y.current)).
> +
> + 2 Remarks: The expression in the noexcept clause is equivalent to 
> +   noexcept(ranges::iter_swap(declval<I>(), declval<I>())) && noexcept(--declval<I&>()).
> ```

In [move.iterator], add declarations to the synopsis:
> ```diff
>   reference operator[](difference_type n) const
>     requires RandomAccessIterator<I>();
>
> + friend constexpr rvalue_reference_t<I> iter_move(const move_iterator& i)
> +   noexcept(see below);
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const move_iterator& x, const move_iterator<I2>& y)
> +     noexcept(see below);
> +
> private:
>   I current; // exposition only
> };
> ```

Add new paragraphs to [move.iter.nonmember] as follows:

> ```diff
> template <RandomAccessIterator I>
>   move_iterator<I>
>     operator+(
>       difference_type_t<I> n,
>       const move_iterator<I>& x);
>
> 2 Effects: Equivalent to x + n.
> +
> + friend constexpr rvalue_reference_t<I> iter_move(const move_iterator& i)
> +   noexcept(see below):
> +
> + -?- Effects: Equivalent to: return ranges::iter_move(i.current);
> +
> + -?- Remarks: The expression in the noexcept clause is equivalent to 
> +     noexcept(ranges::iter_move(i.current)).
> +
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const move_iterator& x, const move_iterator<I2>& y)
> +     noexcept(see below):
> +
> + -?- Effects: Equivalent to ranges::iter_swap(x.current, y.current).
> +
> + -?- Remarks: The expression in the noexcept clause is equivalent to 
> +     noexcept(ranges::iter_swap(x.current, y.current)).
>
> template <InputIterator I>
>   move_iterator<I> make_move_iterator(I i);
>
> 3 Returns: move_iterator<I>(i).
> ```

In [common.iterator], add declarations to the synopsis:

> ```diff
>   common_iterator& operator++();
>   common_iterator operator++(int);
> +
> + friend constexpr rvalue_reference_t<I> iter_move(const common_iterator& i)
> +   noexcept(see below)
> +     requires InputIterator<I>();
> + template <IndirectlySwappable<I> I2, class S2>
> +   friend void iter_swap(const common_iterator& x, const common_iterator<I2, S2>& y)
> +     noexcept(see below)
>
> private:
>   bool is_sentinel; // exposition only
> ```

Add new sections after [common.iter.op.comp]:

> ```diff
> + 24.7.4.2.? iter_move [common.iter.op.iter_move]
> +
> + friend constexpr rvalue_reference_t<I> iter_move(const common_iterator& i)
> +   noexcept(see below)
> +     requires InputIterator<I>();
> +
> + 1 Requires: !i.is_sentinel.
> +
> + 2 Effects: Equivalent to: return ranges::iter_move(i.iter);
> +
> + 3 Remarks: The expression in the noexcept clause is equivalent to
> +   noexcept(ranges::iter_move(i.iter)).
> +
> + 24.7.4.2.? iter_swap [common.iter.op.iter_swap]
> +
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const common_iterator& x, const common_iterator<I2>& y)
> +     noexcept(see below);
> +
> + 1 Requires: !x.is_sentinel && !y.is_sentinel.
> +
> + 2 Effects: Equivalent to ranges::iter_swap(x.iter, y.iter).
> +
> + 3 Remarks: The expression in the noexcept clause is equivalent to 
> +   noexcept(ranges::iter_swap(x.iter, y.iter)).
> ```

In [counted.iterator], add declarations to the synopsis:

> ```diff
>   see below operator[](difference_type n) const
>     requires RandomAccessIterator<I>();
>
> + friend constexpr rvalue_reference_t<I> iter_move(const counted_iterator& i)
> +   noexcept(see below)
> +     requires InputIterator<I>();
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const counted_iterator& x, const counted_iterator<I2>& y)
> +     noexcept(see below);
> +
> private:
>   I current; // exposition only
> ```

Add new paragraphs to [counted.iter.nonmember] as follows:

> ```diff
> template <RandomAccessIterator I>
>   counted_iterator<I>
>     operator+(difference_type_t<I> n, const counted_iterator<I>& x);
>
> 5 Requires: n <= x.cnt.
>
> 6 Effects: Equivalent to x + n.
> +
> + friend constexpr rvalue_reference_t<I> iter_move(const counted_iterator& i)
> +   noexcept(see below)
> +     requires InputIterator<I>();
> +
> + -?- Effects: Equivalent to: return ranges::iter_move(i.current).
> +
> + -?- Remarks: The expression in the noexcept clause is equivalent to 
> +      noexcept(ranges::iter_move(i.current)).
> +
> + template <IndirectlySwappable<I> I2>
> +   friend void iter_swap(const counted_iterator& x, const counted_iterator<I2>& y)
> +     noexcept(see below);
> +
> + -?- Effects: Equivalent to ranges::iter_swap(x.current, y.current).
> +
> + -?- Remarks: The expression in the noexcept clause is equivalent to
> +     noexcept(ranges::iter_swap(x.current, y.current)).
>
> template <Iterator I>
>   counted_iterator<I> make_counted_iterator(I i, difference_type_t<I> n);
>
> 7 Requires: n >= 0.
>
> 8 Returns: counted_iterator<I>(i, n).
> ```

### PR DESCRIPTION
**Pull request by CaseyCarter**
…_move and iter_swap

Fixes #245.

### CODE REVIEW

File: iterators.tex [hunk 1]
```diff
@@ -3591,6 +3644,43 @@
 \tcode{return x + n;}
 \end{itemdescr}
 
+\indexlibrary{\idxcode{iter_move}!\idxcode{move_iterator}}%
```
Comment (ericniebler) [inline iterators.tex]: Do we need a `\rSec4[move.iter.iter_move]{\tcode{iter_move}}` here?

File: iterators.tex [hunk 2]
```diff
@@ -3591,6 +3644,43 @@
...
+\pnum
+\effects Equivalent to: \tcode{return ranges::iter_move(i.current);}
+
+\pnum
+\remarks The expression in \tcode{noexcept} is equivalent to:
+\begin{codeblock}
+noexcept(ranges::iter_move(i.current))
+\end{codeblock}
+\end{itemdescr}
+
```
Comment (ericniebler) [inline iterators.tex]: `\rSec4[move.iter.iter_swap]{\tcode{iter_swap}}` ?

File: iterators.tex [hunk 3]
```diff
@@ -4188,6 +4285,55 @@
...
+  friend void iter_swap(const common_iterator& x, const common_iterator<I2>& y)
+    noexcept(@\seebelow@);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\requires \tcode{!x.is_sentinel \&\& !y.is_sentinel}.
+
+\pnum
+\effects Equivalent to \tcode{ranges::iter_swap(x.iter, y.iter)}.
```
Comment (ericniebler) [inline iterators.tex]: Do we need a colon after "Equivalent to"?

File: iterators.tex [hunk 4]
```diff
@@ -4841,6 +4995,44 @@
 \tcode{return x + n;}
 \end{itemdescr}
 
+\indexlibrary{\idxcode{iter_move}!\idxcode{counted_iterator}}%
```
Comment (ericniebler) [inline iterators.tex]: Needs an `\rSec4`, yes?

File: iterators.tex [hunk 5]
```diff
@@ -4841,6 +4995,44 @@
...
+\effects Equivalent to: \tcode{return ranges::iter_move(i.current);}
+
+\pnum
+\remarks The expression in \tcode{noexcept} is equivalent to:
+\begin{codeblock}
+noexcept(ranges::iter_move(i.current))
+\end{codeblock}
+\end{itemdescr}
+
+\indexlibrary{\idxcode{iter_swap}!\idxcode{counted_iterator}}%
```
Comment (ericniebler) [inline iterators.tex]: Needs an `\rSec4`, yes?

File: iterators.tex (hunk 1 above)
Comment (CaseyCarter) [inline iterators.tex]: We have two styles of presentation for iterator adapters: subclause-per-function, and subclause-per-grouping. This subclause is a grouping of "non-member functions", so `iter_move` and `iter_swap` fit right in here without a distinct subsection heading.

File: iterators.tex (hunk 2 above)
Comment (CaseyCarter) [inline iterators.tex]: Ditto nope.

File: iterators.tex (hunk 3 above)
Comment (CaseyCarter) [inline iterators.tex]: The rules are:
* Equivalent to `<expression>`.
* Equivalent to: `<statement>`
* Equivalent to:
  ```c++
  code;
  block;
  ```

File: iterators.tex (hunk 4 above)
Comment (CaseyCarter) [inline iterators.tex]: Nope - same reason:  this is a "non-member functions" subclause.

File: iterators.tex (hunk 5 above)
Comment (CaseyCarter) [inline iterators.tex]: Ditto nope.

File: iterators.tex (hunk 3 above)
Comment (CaseyCarter) [inline iterators.tex]: (So no, we do not.)

**Comment by CaseyCarter** [comment]
All of these are unfixable (non-editorial bugs in `reverse_iterator` for  which you've already filed an issue) or by design.

**Pull request merged**